File: stratego/stratego-0.4.20/spec/slib/spec/io.r
\literate[{\btt IO}: Input and Output]

% Copyright (C) 1998, 1999, 2000 Eelco Visser <visser@acm.org>
% 
% This program is free software; you can redistribute it and/or modify
% it under the terms of the GNU General Public License as published by
% the Free Software Foundation; either version 2, or (at your option)
% any later version.
% 
% This program is distributed in the hope that it will be useful,
% but WITHOUT ANY WARRANTY; without even the implied warranty of
% MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
% GNU General Public License for more details.
% 
% You should have received a copy of the GNU General Public License
% along with this program; if not, write to the Free Software
% Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
% 02111-1307, USA.

	\begin{abstract}

	A transformation system needs to input terms to transform
	and output transformed terms. Instead of providing a single
	fixed mechanism for IO, this module
	defines primitives for file input and output of
	terms and strings. These primitives can be used in a variety
	of ways to define customized IO.

	\end{abstract}

	A compiled Stratego specification applies the strategy
	\verb|main| to the command line options that it gets. When
	interpreting these it will probably be necessary to read in
	a term from file and later write the transformed term back
	to (another) file. This module provides the primitives for
	doing file input and output. Module \verb|options| defines
	strategies to parse and analyze the command line options.

\begin{code}
module io
imports list string tuple integers time exec
\end{code}

	A \verb|file| can be a string or one of the terms
	\verb|stdin|, \verb|stdout|, \verb|stderr|.

\begin{code}
signature
  sorts File
  constructors
    stdin  : File
    stdout : File
    stderr : File
\end{code}

	\verb|<ReadFromFile> file| reads the term in \verb|file|. The
	\verb|file| needs to be in textual or binary ATerm format.

	\verb|<WriteToTextFile> (file, term)| writes \verb|term| to
	file in textual ATerm format.

	\verb|<WriteToBinaryFile> (file, term)| writes \verb|term| to
	file in BAF format.

 	\verb|<print> (file, [t1,...,tn])| prints the terms \verb|ti|
	to file. If \verb|ti| is a string it is printed without quotes,
	otherwise it is printed as a term. \verb|printnl| has the same
	behaviour, but also prints a newline after \verb|tn|.

	Before printing to a file the file should be opened using
	\verb|<open-file> filename|, which truncates the file, or creates
	it if it doesn't exist.  To append to a file, open the file
	with \verb|<append-file> filename|.  The file is created if it
	doesn't exist.

\begin{code}
strategies

  print             = prim("_ST_print")
  printnl           = prim("_ST_printnl")
  printascii        = prim("_ST_printascii")

  file-exists	    = prim("_ST_file_exists")
  open-file	    = prim("_ST_open_file")
  append-file	    = prim("_ST_append_file")
  close-file	    = prim("_ST_close_file")

  ReadFromFile      = prim("_ST_ReadFromFile")
  WriteToBinaryFile = prim("_ST_WriteToBinaryFile")
  WriteToTextFile   = prim("_ST_WriteToTextFile")
\end{code}

 	The primitive \verb|print-stack| prints the top n elements of
 	the stack if applied as \verb|<print-stack>| n or the entire
	stack if
 	applied to a non-integer term.

\begin{code}
  print-stack       = prim("_ST_PrintStack")
\end{code}

	The strategy \verb|debug| prints the
	current term to \verb|stderr| without changing it. This is
	a useful strategy for debugging specifications (hence its
	name).

\begin{code}
strategies

  open(file) = file; ReadFromFile

  save(file) = split(file, id); WriteToTextFile

  debug      = where(split(!stderr, \x -> [x]\); printnl)
  debug(msg) = where(split(!stderr, \x -> [<msg>(),x]\); printnl)

  say(msg) = where(msg; debug)

  debug-stdout(msg) = where(split(!stdout, \x -> [<msg>(),x]\); printnl)

  trace(msg,s) =
    debug(msg); (s; debug(!"succeeded: ") <+ debug(!"failed: "))

  error = where(split(!stderr, id); printnl)

  fatal-error = where(error; <exit> 1)

  printchar   = where(split(!stdout, \x -> [x]\ ); printascii)

  printstring = where(split(!stdout, \x -> [x]\ ); print)

  print-strings-nl(out) = where(split(out, id); printnl)

  obsolete(msg) = where(<debug(msg)> ": obsolete library strategy")
\end{code}

	The operator \verb|stdio| implements a simple user-interface
	for transformers. A term is read from standard input,
	transformed with parameter strategy \verb|s| and then written
	to standard output. If the transformation failed the text
	\verb|rewriting failed| is written to standard error.

\begin{code}
strategies

  stdio(s) = <ReadFromFile> stdin;
             s; 
             split(!stdout, id); WriteToTextFile
          <+ <fatal-error> ["** rewriting failed"]
\end{code}

	A variant of this strategy provides a pair of the command-line
	options and the input file to the strategy.

\begin{code}
strategies

  stdioO(s) = split(id, <ReadFromFile> stdin); 
              s;
              split(!stdout, id); WriteToTextFile
           <+ <fatal-error> ["** rewriting failed"]
\end{code}

File: stratego/stratego-0.4.18/spec/bak.slib/src/io.c
/*

Copyright (C) 1998, 1999 Eelco Visser <visser@acm.org>

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2, or (at your option)
any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
02111-1307, USA.

*/

/* $Id: io.c,v 1.10 2000/06/04 14:49:40 visser Exp $ */

#include "stratego.h"
#include "tuples.h"
#include <stdlib.h>

/* IO */

ATermTable ST_file_table = NULL;

void ST_file_table_init(void)
{
  ST_file_table = ATtableCreate(16, 80);
  ATtablePut(ST_file_table, ATmake("stdin"),  ATmake("<int>", (int)stdin));
  ATtablePut(ST_file_table, ATmake("stdout"), ATmake("<int>", (int)stdout));
  ATtablePut(ST_file_table, ATmake("stderr"), ATmake("<int>", (int)stderr));
}

ATbool ST_file_exists(void)
{
  char *name;
  FILE *file;
  name = t_string(Ttop);

  if((file = fopen(name, "r")) == NULL)
    return ATfalse;
  else
    { 
      fclose(file);
      return ATtrue;
    }
}

ATbool ST_open_file(void)
{
  FILE *file;
  ATerm name;

  if(ST_file_table == NULL)
    ST_file_table_init();
  
  name = Ttop;
  if(ATmatch(name, "stdout"))
    file = stdout;
  else if(ATmatch(name, "stderr"))
    file = stderr;
  else if(ATmatch(name, "stdin"))
    file = stdin;
  else if(t_is_string(name))
    {
      file = fopen(t_string(name), "w");
      if(file == NULL)
	return ATfalse;
      ATtablePut(ST_file_table, name, ATmake("<int>", (int)file));      
    }
  return ATtrue;
}

ATbool ST_append_file(void)
{
  FILE *file;
  ATerm name;

  if(ST_file_table == NULL)
    ST_file_table_init();
  
  name = Ttop;
  if(ATmatch(name, "stdout"))
    file = stdout;
  else if(ATmatch(name, "stderr"))
    file = stderr;
  else if(ATmatch(name, "stdin"))
    file = stdin;
  else if(t_is_string(name))
    {
      file = fopen(t_string(name), "a");
      if(file == NULL)
	return ATfalse;
      ATtablePut(ST_file_table, name, ATmake("<int>", (int)file));      
    }
  return ATtrue;
}

FILE *_ST_file_table_lookup(ATerm name)
{
  ATerm desc;

  /* ATfprintf(stderr, "_ST_file_table_lookup(%t)\n", name); */

  if(ST_file_table == NULL)
    ST_file_table_init();
  
  if((desc = ATtableGet(ST_file_table, name)) == NULL)
    return NULL;
  else
    return (FILE *)ATgetInt((ATermInt)desc);
}

ATbool ST_close_file(void)
{
  FILE *file;
  ATerm name;

  if(ST_file_table == NULL)
    ST_file_table_init();
  
  name = Ttop;
  if(ATmatch(name, "stdout"))
    fclose(stdout);
  else if(ATmatch(name, "stderr"))
    fclose(stderr);
  else if(ATmatch(name, "stdin"))
    fclose(stdin);
  else if((file = _ST_file_table_lookup(name)) != NULL)
    fclose(file);
  else
    return ATfalse;

  return ATtrue;
}

ATbool ST_print(void)
{  
  FILE *outfile;
  ATerm file, str;

  if(MatchPair(Ttop, &file, &str))
    {
      outfile = _ST_file_table_lookup(file);
      if(outfile == NULL) 
	return ATfalse;
      str = consnil_to_list(str);
      while(!ATisEmpty((ATermList)str))
	{
	  if(ATisString(ATgetFirst((ATermList)str)))
	    ATfprintf(outfile, "%s", t_string(ATgetFirst((ATermList)str)));
	  else if(ATisReal(ATgetFirst((ATermList)str)))
	    ATfprintf(outfile, "%.2f", ATgetReal((ATermReal)ATgetFirst((ATermList)str)));
	  else
	    ATfprintf(outfile, "%t", ATgetFirst((ATermList)str));
	  str = (ATerm)ATgetNext((ATermList)str);
	}
      return ATtrue;
    }
  return ATfalse; 
}

ATbool ST_printnl(void)
{  
  FILE *outfile;
  ATerm file, str;

  if(MatchPair(Ttop, &file, &str))
    {
      outfile = _ST_file_table_lookup(file);
      if(outfile == NULL) {
	ATfprintf(stderr, "printnl: could not open file: %t\n", file);
	return ATfalse;
      }
      str = consnil_to_list(str);
      if(!(ATgetType(str) == AT_LIST)) {
	ATfprintf(stderr, "ST_printnl: str not a list: %t\n", str);
	return ATfalse;
      }
      while(!ATisEmpty((ATermList)str))
	{
	  if(ATisString(ATgetFirst((ATermList)str)))
	    ATfprintf(outfile, "%s", t_string(ATgetFirst((ATermList)str)));
	  else if(ATisReal(ATgetFirst((ATermList)str)))
	    ATfprintf(outfile, "%.2f", ATgetReal((ATermReal)ATgetFirst((ATermList)str)));
	  else
	    ATfprintf(outfile, "%t", ATgetFirst((ATermList)str));
	  str = (ATerm)ATgetNext((ATermList)str);
	}
      ATfprintf(outfile, "\n");
      return ATtrue;
    }
  return ATfalse; 
}

ATbool ST_printascii(void)
{   
  FILE *outfile;
  ATerm file, str;

  /* ATfprintf(stderr, "ST_printascii(): top = %t\n", Ttop); */

  if(MatchPair(Ttop, &file, &str))
    {
      outfile = _ST_file_table_lookup(file);
      if(outfile == NULL) 
	return ATfalse;
      str = consnil_to_list(str);
      while(!ATisEmpty((ATermList)str))
	{
	  if(ATisInt(ATgetFirst((ATermList)str)))
	    ATfprintf(outfile, "%c", 
		      ATgetInt((ATermInt)ATgetFirst((ATermList)str)));
	  else
	    ATfprintf(outfile, "%t", ATgetFirst((ATermList)str));
	  str = (ATerm)ATgetNext((ATermList)str);
	}
      /* ATfprintf(outfile, "\n"); */
      return ATtrue;
    }
  return ATfalse; 
}

ATbool ST_ReadFromFile(void)
{
  ATerm filename, in_term;
  FILE *infile;

  filename = Ttop;

  if(ATmatch(filename, "stdin"))
    infile = stdin;
  else if(ATisString(filename))
    infile = fopen(t_string(filename), "r");
  else
    return ATfalse;

  if((in_term = ATreadFromFile(infile)) == NULL)
    {
      ATfprintf(stderr, "not a valid term\n");
      return ATfalse;
    }

  fclose(infile);

  /* ATfprintf(stderr, "in_term = %t\n", in_term); */
  
  Tset(list_to_consnil(in_term));
  return ATtrue;
}

ATbool ST_WriteToFile(ATbool binary)
{  
  FILE *outfile;
  ATerm filename, trm;
  if(MatchPair(Ttop, &filename, &trm))
    {
      if(ATmatch(filename, "stdout"))
	outfile = stdout;
      else if(ATmatch(filename, "stderr"))
	outfile = stderr;
      else if(ATisString(filename))
	outfile = fopen(t_string(filename), "w");
      else return ATfalse;
      if(outfile != NULL)
	{
	  if(binary)
	    ATwriteToBinaryFile(consnil_to_list(trm), outfile);
	  else 
	    {
	      ATwriteToTextFile(consnil_to_list(trm), outfile);
	      fprintf(outfile, "\n");
	    }
	  fclose(outfile);
	}
      return ATtrue;
    }
  return ATfalse; 
}

ATbool ST_WriteToBinaryFile(void)
{
  return ST_WriteToFile(ATtrue);
}


ATbool ST_WriteToTextFile(void)
{
  return ST_WriteToFile(ATfalse);
}

ATbool ST_PrintStack(void)
{
  ATerm t;
  t = Ttop;
  if(ATisInt(t)) 
    TprintStackTop(ATgetInt((ATermInt)t), "");
  else
    TprintStack();    
  return ATtrue;
}
